File: 0055-jump-game/0055-jump-game.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var canJump = function (nums) {
    let n = nums.length;
    let dp = new Array(n).fill(false);

    dp[n - 1] = true;

    for (let i = n - 2; i >= 0; i--) {
        for (let j = i + 1; j <= i + nums[i]; j++) {
            if (dp[j]) {
                dp[i] = true;
                break;
            }
        }
    }

    return dp[0];
};

File: 0055-jump-game/0055-jump-game.py
from functools import lru_cache


class Solution:
    def canJump(self, nums: list[int]) -> bool:
        n = len(nums)

        @lru_cache(maxsize=None)
        def solve(i):
            if i == n - 1:
                return True

            maxJump = min(i + nums[i], n - 1)
            for pos in range(i + 1, maxJump + 1):
                if solve(pos):
                    return True

            return False

        return solve(0)
